Multi-user feature code fails to compile when cross compiling for ARMv7

Krill version tested:
```
$ rustc -V
rustc 1.57.0 (f1edd0429 2021-11-29)

$ git describe --all --long
heads/main-0-g3ab1f38c

$ lsb_release -d
Description:    Ubuntu 21.10

$ uname -a
Linux ximon-laptop 5.15.0-051500-generic #202110312130 SMP Sun Oct 31 21:33:20 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

Prerequisites:

```
$ cargo install cross
$ rustup target add armv7-unknown-linux-gnueabihf
```

Without the `multi-user` feature (which is normally enabled by default):

```
$ cross build --target armv7-unknown-linux-gnueabihf --no-default-features --features static-openssl
    Finished dev [unoptimized + debuginfo] target(s) in 2m 25s
```

With the `multi-user` feature (i.e. the default):

```
$ cross build --target armv7-unknown-linux-gnueabihf --features static-openssl
...
error[E0277]: the trait bound `AtomicU64: Deserialize<'_>` is not satisfied
    --> src/daemon/auth/common/crypt.rs:44:5
     |
44   |     counter: AtomicU64,     // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
     |     ^^^^^^^ the trait `Deserialize<'_>` is not implemented for `AtomicU64`
     |
note: required by `next_element`
```

However, without cross compiling it works:
```
$ cargo build --features static-openssl
...
    Finished dev [unoptimized + debuginfo] target(s) in 3m 13s
```

--- a/src/commons/util/ext_serde.rs
+++ b/src/commons/util/ext_serde.rs
@@ -1,5 +1,5 @@
 //! Defines helper methods for Serializing and Deserializing external types.
-use std::str::FromStr;
+use std::{str::FromStr, sync::atomic::{AtomicU64, Ordering}};
 
 use bytes::Bytes;
 use log::LevelFilter;
@@ -161,4 +161,23 @@ where
 {
     let string = String::deserialize(d)?;
     Facility::from_str(&string).map_err(|_| de::Error::custom(format!("Unsupported syslog_facility: \"{}\"", string)))
+}
+
+//------------- AtomicU64 -----------------------------------------------------
+// Implemented automatically by Serde derive but only for x86_64 architectures,
+// for other architectures (such as armv7 for the Raspberry Pi 4b) it has to be
+// implemented manually.
+
+pub fn de_atomicu64<'de, D>(d: D) -> Result<AtomicU64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Ok(AtomicU64::new(u64::deserialize(d)?))
+}
+
+pub fn ser_atomicu64<S>(v: &AtomicU64, s: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    s.serialize_u64(v.load(Ordering::SeqCst))
 }

--- a/src/daemon/auth/common/crypt.rs
+++ b/src/daemon/auth/common/crypt.rs
@@ -38,61 +38,10 @@ const POLY1305_TAG_BYTE_LEN: usize = POLY1305_TAG_BIT_LEN / 8;
 const CLEARTEXT_PREFIX_LEN: usize = CHACHA20_NONCE_BYTE_LEN + POLY1305_TAG_BYTE_LEN;
 const UNUSED_AAD: [u8; 0] = [0; 0];
 
-//------------ CrossPlatformDeserializableAtomicU64 -------------------------------------------------------------------
-
-/// This type provides Serde (de)serialization irrespective of target platform, as the Serde crate provided derive
-/// impls only appear to produce code when compiled on x86_64. See: https://github.com/NLnetLabs/krill/issues/751
-#[derive(Debug)]
-struct CrossPlatformDeserializableAtomicU64(AtomicU64);
-
-impl std::ops::Deref for CrossPlatformDeserializableAtomicU64 {
-    type Target = AtomicU64;
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
-
-impl<'de> serde::Deserialize<'de> for CrossPlatformDeserializableAtomicU64 {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        struct DeserializableAtomicU64Visitor;
-
-        impl<'de> serde::de::Visitor<'de> for DeserializableAtomicU64Visitor {
-            type Value = CrossPlatformDeserializableAtomicU64;
-
-            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
-                formatter.write_str("an unsigned integer in the range 0..=2^64-1")
-            }
-
-            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
-            where
-                E: serde::de::Error, {
-                Ok(CrossPlatformDeserializableAtomicU64(AtomicU64::new(v)))
-            }
-        }
-
-        deserializer.deserialize_u64(DeserializableAtomicU64Visitor)
-    }
-}
-
-impl serde::Serialize for CrossPlatformDeserializableAtomicU64 {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer,
-    {
-        serializer.serialize_u64(self.0.load(Ordering::SeqCst))
-    }
-}
-
-//------------ NonceState ---------------------------------------------------------------------------------------------
-
 #[derive(Debug, Deserialize, Serialize)]
 pub struct NonceState {
-    sender_unique: [u8; 4],           //   32 bits
-    counter: CrossPlatformDeserializableAtomicU64, // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
+    sender_unique: [u8; 4], //   32 bits
+    counter: AtomicU64,     // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
 }
 
 impl NonceState {
@@ -103,7 +52,7 @@ impl NonceState {
 
         Ok(NonceState {
             sender_unique,
-            counter: CrossPlatformDeserializableAtomicU64(AtomicU64::new(0)),
+            counter: AtomicU64::new(0),
         })
     }
 
@@ -119,8 +68,6 @@ impl NonceState {
         nonce
     }
 }
-
-//------------ CryptState ---------------------------------------------------------------------------------------------
 
 pub struct CryptState {
     pub key: [u8; CHACHA20_KEY_BYTE_LEN],

--- a/src/daemon/auth/common/crypt.rs
+++ b/src/daemon/auth/common/crypt.rs
@@ -38,10 +38,61 @@ const POLY1305_TAG_BYTE_LEN: usize = POLY1305_TAG_BIT_LEN / 8;
 const CLEARTEXT_PREFIX_LEN: usize = CHACHA20_NONCE_BYTE_LEN + POLY1305_TAG_BYTE_LEN;
 const UNUSED_AAD: [u8; 0] = [0; 0];
 
+//------------ CrossPlatformDeserializableAtomicU64 -------------------------------------------------------------------
+
+/// This type provides Serde (de)serialization irrespective of target platform, as the Serde crate provided derive
+/// impls only appear to produce code when compiled on x86_64. See: https://github.com/NLnetLabs/krill/issues/751
+#[derive(Debug)]
+struct CrossPlatformDeserializableAtomicU64(AtomicU64);
+
+impl std::ops::Deref for CrossPlatformDeserializableAtomicU64 {
+    type Target = AtomicU64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for CrossPlatformDeserializableAtomicU64 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct DeserializableAtomicU64Visitor;
+
+        impl<'de> serde::de::Visitor<'de> for DeserializableAtomicU64Visitor {
+            type Value = CrossPlatformDeserializableAtomicU64;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("an unsigned integer in the range 0..=2^64-1")
+            }
+
+            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error, {
+                Ok(CrossPlatformDeserializableAtomicU64(AtomicU64::new(v)))
+            }
+        }
+
+        deserializer.deserialize_u64(DeserializableAtomicU64Visitor)
+    }
+}
+
+impl serde::Serialize for CrossPlatformDeserializableAtomicU64 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u64(self.0.load(Ordering::SeqCst))
+    }
+}
+
+//------------ NonceState ---------------------------------------------------------------------------------------------
+
 #[derive(Debug, Deserialize, Serialize)]
 pub struct NonceState {
-    sender_unique: [u8; 4], //   32 bits
-    counter: AtomicU64,     // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
+    sender_unique: [u8; 4],           //   32 bits
+    counter: CrossPlatformDeserializableAtomicU64, // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
 }
 
 impl NonceState {
@@ -52,7 +103,7 @@ impl NonceState {
 
         Ok(NonceState {
             sender_unique,
-            counter: AtomicU64::new(0),
+            counter: CrossPlatformDeserializableAtomicU64(AtomicU64::new(0)),
         })
     }
 
@@ -68,6 +119,8 @@ impl NonceState {
         nonce
     }
 }
+
+//------------ CryptState ---------------------------------------------------------------------------------------------
 
 pub struct CryptState {
     pub key: [u8; CHACHA20_KEY_BYTE_LEN],

--- a/src/daemon/auth/common/crypt.rs
+++ b/src/daemon/auth/common/crypt.rs
@@ -27,6 +27,7 @@ use std::{
 use crate::commons::{
     error::{Error, KrillIoError},
     KrillResult,
+    util::ext_serde,
 };
 
 const CHACHA20_KEY_BIT_LEN: usize = 256;
@@ -41,6 +42,11 @@ const UNUSED_AAD: [u8; 0] = [0; 0];
 #[derive(Debug, Deserialize, Serialize)]
 pub struct NonceState {
     sender_unique: [u8; 4], //   32 bits
+
+    #[serde(
+        deserialize_with = "ext_serde::de_atomicu64",
+        serialize_with = "ext_serde::ser_atomicu64"
+    )]
     counter: AtomicU64,     // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
 }
 

